wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/JpaWorkspaceActivityDao.java

```diff
@@ -38,71 +40,190 @@
 
   @Override
   public void setExpiration(WorkspaceExpiration expiration) throws ServerException {
-    requireNonNull(expiration, "Required non-null expiration object");
+    setExpirationTime(expiration.getWorkspaceId(), expiration.getExpiration());
+  }
+
+  @Override
+  public void setExpirationTime(String workspaceId, long expirationTime) throws ServerException {
+    requireNonNull(workspaceId, "Required non-null workspace id");
+    withActivity(workspaceId, a -> a.setExpiration(expirationTime));
+  }
+
+  @Override
+  public void removeExpiration(String workspaceId) throws ServerException {
+    requireNonNull(workspaceId, "Required non-null workspace id");
+    withOptionalActivity(workspaceId, a -> a.setExpiration(null));
+  }
+
+  @Override
+  public List<String> findExpired(long timestamp) throws ServerException {
     try {
-      doCreateOrUpdate(expiration);
+      return doFindExpired(timestamp);
     } catch (RuntimeException x) {
       throw new ServerException(x.getLocalizedMessage(), x);
     }
   }
 
   @Override
-  public void removeExpiration(String workspaceId) throws ServerException {
-    requireNonNull(workspaceId, "Required non-null id");
+  @Transactional
+  public void removeActivity(String workspaceId) throws ServerException {
+    EntityManager em = managerProvider.get();
+    WorkspaceActivity activity = em.find(WorkspaceActivity.class, workspaceId);
     try {
-      doRemove(workspaceId);
+      if (activity != null) {
+        em.remove(activity);
+        em.flush();
+      }
     } catch (RuntimeException x) {
       throw new ServerException(x.getLocalizedMessage(), x);
     }
   }
 
   @Override
-  public List<String> findExpired(long timestamp) throws ServerException {
+  public void setCreatedTime(String workspaceId, long createdTimestamp) throws ServerException {
```

Comment:
I guess it would be better to have Transactional here.

Reply:
the `doUpdate()` method that does all the real work is transactional already. Do you think it makes sense to have the public methods `@Transactional` instead of the protected helpers?

Frankly, I was just following suite to conform how the class was implemented already, but IMHO it would be better to annotate only the API methods to have an explicit picture of their behavior.

Reply:
> the doUpdate() method that does all the real work is transactional already. 

sorry, it's my fault, maybe I was taking a look in-memory impl and commenting `JPA`.

> Do you think it makes sense to have the public methods @Transactional instead of the protected helpers?

No, I don't think so. Current approach is OK for me